### PR TITLE
Desktop: Assign note to Notebook or Sub_Notebook functionality added

### PR DESCRIPTION
**Pull request by coderrsid**
Fixed Issue #2296 

This pull request add the functionality of assigning note to another notebook.

# WORKING
GIF

### CODE REVIEW

**Comment by coderrsid** [comment]
Special thanks to @tessus for helping me out for every query. I seek to participate in Google Summer of Code and work with Joplin.

**Comment by laurent22** [comment]
The animated gif doesn't seem to work. How does your code handle sub notebooks?

**Comment by tessus** [comment]
I'm not sure but the animated gif only shows how a note is created. This PR is for moving one or more notes to an existing folder.

**Comment by coderrsid** [comment]
> The animated gif doesn't seem to work. How does your code handle sub notebooks?

Hey @laurent22, Thanks for asking. I just updated the GIF, actually it was not the whole GIF earlier. Now you can see we can assign note to any notebook or sub notebook.

**Comment by coderrsid** [comment]
> I'm not sure but the animated gif only shows how a note is created. This PR is for moving one or more notes to an existing folder.
Hey @tessus, actually it was not the whole GIF earlier. Now you can see we can assign note to any notebook or sub notebook.

**Comment by tessus** [comment]
Thanks for the new gif.

I have 2 questions:

- how is the search implemented? e.g. if you start with a, does it show all the folders that start wit a or that contain a?
- the folder structure is flat. Is there any chance that sub folders are indented relative to their parents?

**Comment by tessus** [comment]
There must be a code already available that does the folder list with indentions. The web clipper allows you to choose a folder from a list which has the proper structure. It's also available on the mobile client, when you move a note to a different folder. These 2 just don't allow you to search.

But maybe the indention is just a stupid idea, but I like it when I have a structure.

**Comment by coderrsid** [comment]
> There must be a code already available that does the folder list with indentions. The web clipper allows you to choose a folder from a list which has the proper structure. It's also available on the mobile client, when you move a note to a different folder. These 2 just don't allow you to search.
> 
> But maybe the indention is just a stupid idea, but I like it when I have a structure.

Search implementation is already done @tessus . If you want the subfolders indented too? I can work on that.
GIF2

**Comment by tessus** [comment]
Ok, I can see it now. It's just a bit confusing, because in the beginning, when there's just an `a` in the search field, it shows all folders. This should not be the case.
But I was asking how the search algorithm works. But I think I figured it from the gif. It lists all folders that contain the search string (and not start with it).

> If you want the subfolders indented too?

Let's wait for Laurent's input.

**Comment by mic704b** [comment]
Hi @coderrsid the text in the video is small. Is it possible to zoom into the relevant part of screen to show clearly what is happening?

**Comment by coderrsid** [comment]
@tessus. It's the defaultValue.. which i have set to the folder[0] (first folder in the list), I can change it to empty value too

**Comment by tessus** [comment]
Yep, I think it should be empty by default, in which case it should show all folders (with indention). As soon as you start typing only the results are shown. And they don't have to be indented.
What do you think? Does this make sense?

**Comment by tessus** [comment]
When I come to think of it, we have missed something: Folders can have the same name.

Take the following structure:

```
A
   B
   C
D
   B
   E
F
   G
   H
   I
      B
```

When you search for `B`, you will get three `B` in the result set:

```
B
B
B
```

Even with indention, we can't solve this problem.

**Comment by laurent22** [comment]
> If you want the subfolders indented too?

Yes it should be indented. Please have a look at how it's done in the clipper for example.

**Comment by coderrsid** [comment]
> When I come to think of it, we have missed something: Folders can have the same name.
> 
> Take the following structure:
> 
> ```
> A
>    B
>    C
> D
>    B
>    E
> F
>    G
>    H
>    I
>       B
> ```
> 
> When you search for `B`, you will get three `B` in the result set:
> 
> ```
> B
> B
> B
> ```
> 
> Even with indention, we can't solve this problem.

I believe that too.. @tessus .

**Comment by coderrsid** [comment]
> > If you want the subfolders indented too?
> 
> Yes it should be indented. Please have a look at how it's done in the clipper for example.

Yeah i'll have a look. Can you direct me to the clipper source indent files ?

**Comment by coderrsid** [comment]
@laurent22 . I tried changing the code according to the clipper but it seems that the function `nonBreakingSpecify()`.

image

But when i tried to `console.log`, the updated title doesn't have the places as you can compare the code(above) and the console in the (bottom of) screenshot below.
Screenshot 2020-02-24 at 8 03 25 PM

**Comment by tessus** [comment]
> I believe that too.. @tessus .

Yes, but there are a few ideas that will solve the problem.

**Comment by coderrsid** [comment]
> > I believe that too.. @tessus .
> 
> Yes, but there are a few ideas that will solve the problem.

Yeah @tessus. One way it can be solved is by indenting folders like in the clipper @laurent22  suggested by using depth property. We can change the title of the folder in the display by which it shows the name like the image below according to the hierarchy

Screenshot 2020-02-25 at 8 47 51 AM

**Comment by tessus** [comment]
No, I think you do not understand. Look at my example above. When you search for `B` you will get 3 results:

```
B
B
B
```

Even with indention it would look like this:

```
   B
   B
      B
```

So, how exactly does this help?

There are 2 issues here:

1. The part where nothing is in the search field. This is the part where indention makes sense.
2. When you enter a search term and the search result set contains duplicates.

I'm talking about the 2nd item. Indention will do nothing. So you will have to find a solution.
E.g. in case of duplicate enttries in the result set, also show the parents.

```
B (A)
B (D)
B (F <separator> I)
```

Do you understand what I'm trying to say?

**Comment by coderrsid** [comment]
> No, I think you do not understand. Look at my example above. When you search for `B` you will get 3 results:
> 
> ```
> B
> B
> B
> ```
> 
> Even with indention it would look like this:
> 
> ```
>    B
>    B
>       B
> ```
> 
> So, how exactly does this help?
> 
> There are 2 issues here:
> 
> 1. The part where nothing is in the search field. This is the part where indention makes sense.
> 2. When you enter a search term and the search result set contains duplicates.
> 
> I'm talking about the 2nd item. Indention will do nothing. So you will have to find a solution.
> E.g. in case of duplicate enttries in the result set, also show the parents.
> 
> ```
> B (A)
> B (D)
> B (F <separator> I)
> ```
> 
> Do you understand what I'm trying to say?

Yeah yeah i do understand what you're saying instead of parent's name after child.
What we can do instead is that if we search B:

`
B
AB(disabled)
       B
AC(disabled)
       A(disabled)
             B
`

**Comment by tessus** [comment]
Hmm, not a bad idea, but I think you meant:

```
A (disabled)
   B
D (disabled)
   B
F (disabled)
   I (disabled)
      B
```

Yes, this is also a solution. Mine was just an example. And since we most likely won't have too many duplicates anyway, the list should not be too long.

Theoretically you could even use this algorithm for non duplicates. In that case it would be awesome, if we could only jump between the active ones, so that one doesn't have to hit 6x the down key to actually reach the 3rd result.

**Comment by coderrsid** [comment]
Yeah sure we can. That's why i suggested this method. Disabling items will lead to an easy search and will show the parent of sub-notebook too and as you said for moving down in the list too. Can we confirm this from @laurent22 too?

**Comment by tessus** [comment]
Yep, Laurent should definitely chime in.

**Comment by mic704b** [comment]
> if we could only jump between the active ones, so that one doesn't have to hit 6x the down key to actually reach the 3rd result.

Yes, but using the mouse it won't be so much fun.  I think the concise option you first suggested is preferable.

**Comment by coderrsid** [comment]
@PackElend, you can add the label gsoc-2020

**Comment by PackElend** [comment]
@coderrsid I haven't found you in the forum yet, can you tell why you are doing the PR?

**Comment by coderrsid** [comment]
@PackElend. I am in the forum but sorry i hadn't made a formal introduction but now i did. You can find me. Sorry for inconvenience and i do want to participate in GSOC.

**Comment by PackElend** [comment]
>  Sorry for inconvenience and i do want to participate in GSOC.

It git quite late at my side yesterday, so I only copy&paste the question 

no worries, I'm just was wondering to whom we talk to, I found you in the forum but without an introduction. Don't forget to tell us about yourself in the forum and we are fine :)

**Comment by coderrsid** [comment]
> > Sorry for inconvenience and i do want to participate in GSOC.
> 
> It git quite late at my side yesterday, so I only copy&paste the question
> 
> no worries, I'm just was wondering to whom we talk to, I found you in the forum but without an introduction. Don't forget to tell us about yourself in the forum and we are fine :)

Thanks for the information @PackElend. I just introduced myself. 😄

**Comment by coderrsid** [comment]
@laurent22 Can you review the different methods we suggested and discuss your thoughts on it? So we can finalize this pull request features. Thanks

**Comment by laurent22** [comment]
@coderrsid, you've `@mention` me three times in the four minutes. As you can imagine, we're all busy here, and if everybody did like you we'll get hundred of notifications per hour, so please stop and be patient.

**Comment by coderrsid** [comment]
Sure.

File: ElectronClient/gui/MainScreen.jsx [hunk 1]
```diff
@@ -194,6 +194,33 @@ class MainScreenComponent extends React.Component {
...
+					autocomplete: startFolders,
+					onClose: async answer => {
+						if (answer != null) {
+							if (command.noteIds.length === 1) {
+								await Note.moveToFolder(command.noteIds[0], answer.value);
+							} else {
+								for (let i = 0; i < command.noteIds.length; i++) {
+									await Note.moveToFolder(command.noteIds[i], answer.value);
+								}
+							}
```
Comment (laurent22) [inline ElectronClient/gui/MainScreen.jsx]: Please remove the first condition.

File: ElectronClient/app.js [hunk 2]
```diff
@@ -761,7 +761,7 @@ class Application extends BaseApplication {
 					platforms: ['darwin'],
 					accelerator: 'Command+W',
 					selector: 'performClose:',
-				},  {
+				}, {
 					type: 'separator',
 				}, {
 					label: _('Templates'),
```
Comment (laurent22) [inline ElectronClient/app.js]: Please undo these changes.

File: ElectronClient/gui/MainScreen.jsx [hunk 3]
```diff
@@ -157,8 +157,8 @@ class MainScreenComponent extends React.Component {
 				return { value: a.id, label: a.title };
 			})
 				.sort((a, b) => {
-				// sensitivity accent will treat accented characters as differemt
-				// but treats caps as equal
+					// sensitivity accent will treat accented characters as differemt
+					// but treats caps as equal
```
Comment (laurent22) [inline ElectronClient/gui/MainScreen.jsx]: Undo these changes.

File: ElectronClient/gui/MainScreen.jsx [hunk 4]
```diff
@@ -194,6 +194,33 @@ class MainScreenComponent extends React.Component {
 					},
 				},
 			});
+		} else if (command.name === 'setNotebook') {
```
Comment (laurent22) [inline ElectronClient/gui/MainScreen.jsx]: setNotebook => moveToFolder

File: ElectronClient/gui/MainScreen.jsx [hunk 5]
```diff
@@ -194,6 +194,33 @@ class MainScreenComponent extends React.Component {
...
+			const folderItems = this.props.folders;
+			const startFolders = folderItems
+				.map(a => {
+					return { key: a.id, value: a.id, label: a.title };
+				});
+
+			this.setState({
+				promptOptions: {
+					label: _('Assign Notebook:'),
```
Comment (laurent22) [inline ElectronClient/gui/MainScreen.jsx]: Assign Notebook => Move to Notebook

File: ElectronClient/gui/utils/NoteListUtils.js [hunk 6]
```diff
@@ -35,6 +35,19 @@ class NoteListUtils {
 				})
 			);
 
+			menu.append(
+				new MenuItem({
+					label: _('Assign Notebook'),
+					click: async () => {
+						props.dispatch({
+							type: 'WINDOW_COMMAND',
+							name: 'setNotebook',
```
Comment (laurent22) [inline ElectronClient/gui/utils/NoteListUtils.js]: moveToFolder

**Comment by laurent22** [comment]
Regarding the functionalities:

- Please display the notebooks and sub-notebooks as a tree
- Do not pre-populate the dropdown type-in area. Now it defaults to the first notebook in the list, but it should be blank when the dialog open.
- I'm fine with having duplicate notebooks in the list. We can think of a solution later on, but for this PR it's not needed.

File: ElectronClient/gui/MainScreen.jsx (hunk 1 above)
Comment (coderrsid) [reply ElectronClient/gui/MainScreen.jsx]: Done.

File: ElectronClient/app.js (hunk 2 above)
Comment (coderrsid) [reply ElectronClient/app.js]: Done.

File: ElectronClient/gui/MainScreen.jsx (hunk 3 above)
Comment (coderrsid) [reply ElectronClient/gui/MainScreen.jsx]: Done.

File: ElectronClient/gui/MainScreen.jsx (hunk 4 above)
Comment (coderrsid) [reply ElectronClient/gui/MainScreen.jsx]: Done.

File: ElectronClient/gui/MainScreen.jsx (hunk 5 above)
Comment (coderrsid) [reply ElectronClient/gui/MainScreen.jsx]: Done.

File: ElectronClient/gui/utils/NoteListUtils.js (hunk 6 above)
Comment (coderrsid) [reply ElectronClient/gui/utils/NoteListUtils.js]: Done.

**Comment by coderrsid** [comment]
Hey @laurent22 , i tried displaying the notebooks and sub-notebooks as a tree but unfortunately i am not able to do so because of the promptOptions may not be able to show title with blank-spaces, i tried `console logging` too. But however i tried, the code is below, can you please help me and point where am i doing it wrong.

Screenshot 2020-02-27 at 11 55 06 PM

**Comment by tessus** [comment]
Try `\0000a0` instead of `\u00A0`.

**Comment by tessus** [comment]
Also, please don't send screenshots of the code. Why are you doing this? 

You can add code within a code block.

**Comment by Rishgod** [comment]
> Also, please don't send screenshots of the code. Why are you doing this?
> 
> You can add code within a code block.
> 
> Me: "Can you please send me the error message?"
> Somebody sends me: a screenshot of the terminal window, pasted into a Word document, which then was zipped.
> Me: "Aaaaarrrrggghhhh!!!!!"

🤣 😆

**Comment by coderrsid** [comment]
@tessus That code is for octal spaces which are not allowed in strict mode and for the code thing.. i'll remember that.

**Comment by tessus** [comment]
@coderrsid I am a bit puzzled by this. How does the web clipper solve this? Or the tree on `move-to-folder` on mobile? It works there...

**Comment by coderrsid** [comment]
@tessus It's because there is no prompt called there.. only JSX is returned with HTML dropdown (option, select). Below you can see : 
```
const foldersComp = () => {
			const optionComps = [];

			const nonBreakingSpacify = (s) => {
				// https://stackoverflow.com/a/24437562/561309
				return s.replace(/ /g, '\u00a0');
			};

			const addOptions = (folders, depth) => {
				for (let i = 0; i < folders.length; i++) {
					const folder = folders[i];
					optionComps.push(<option key={folder.id} value={folder.id}>{nonBreakingSpacify('    '.repeat(depth) + folder.title)}</option>);
					if (folder.children) addOptions(folder.children, depth + 1);
				}
			};

			addOptions(this.props.folders, 0);

			return (
				<div className="Folders">
					<label>In notebook: </label>
					<select value={this.props.selectedFolderId || ''} onChange={this.folderSelect_change}>
						{ optionComps }
					</select>
				</div>
			);
		};
```

**Comment by tessus** [comment]
@coderrsid so maybe a bug or issue with the dialog prompt then. Hmm...

**Comment by coderrsid** [comment]
@tessus maybe.. although i don't know the deep insights of code. @laurent22 might know about this.

File: ElectronClient/gui/utils/NoteListUtils.js [hunk 7]
```diff
@@ -35,6 +35,19 @@ class NoteListUtils {
 				})
 			);
 
+			menu.append(
+				new MenuItem({
+					label: _('Assign Notebook'),
+					click: async () => {
```
Comment (laurent22) [inline ElectronClient/gui/utils/NoteListUtils.js]: It's not async

File: ElectronClient/gui/utils/NoteListUtils.js [hunk 8]
```diff
@@ -35,6 +35,19 @@ class NoteListUtils {
 				})
 			);
 
+			menu.append(
+				new MenuItem({
+					label: _('Assign Notebook'),
```
Comment (laurent22) [inline ElectronClient/gui/utils/NoteListUtils.js]: "Move to notebook"

File: ElectronClient/gui/MainScreen.jsx [hunk 9]
```diff
@@ -194,6 +194,29 @@ class MainScreenComponent extends React.Component {
...
+		} else if (command.name === 'moveToFolder') {
+			const folderItems = this.props.folders;
+			const startFolders = folderItems
+				.map(a => {
+					return { key: a.id, value: a.id, label: a.title };
+				});
+
+			this.setState({
+				promptOptions: {
+					label: _('Move to Notebook:'),
```
Comment (laurent22) [inline ElectronClient/gui/MainScreen.jsx]: Sorry could you make it lower case? "Move to notebook:" It's for consistency with the rest of the app.

**Comment by laurent22** [comment]
I didn't actually run the code. Is it currently working or is there something missing? The notebook indentation doesn't work?

**Comment by coderrsid** [comment]
@laurent22 I've made the changes. It's still working. The notebook indentation isn't working because of prompt options.

File: ElectronClient/gui/utils/NoteListUtils.js (hunk 7 above)
Comment (coderrsid) [reply ElectronClient/gui/utils/NoteListUtils.js]: Done.

File: ElectronClient/gui/utils/NoteListUtils.js (hunk 8 above)
Comment (coderrsid) [reply ElectronClient/gui/utils/NoteListUtils.js]: Done.

File: ElectronClient/gui/MainScreen.jsx (hunk 9 above)
Comment (coderrsid) [reply ElectronClient/gui/MainScreen.jsx]: Done.

File: ElectronClient/gui/NoteContentPropertiesDialog.js [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+const react_1 = require('react');
+const { _ } = require('lib/locale.js');
+const { themeStyle } = require('../theme.js');
+const DialogButtonRow = require('./DialogButtonRow.min');
+const Countable = require('countable');
+function NoteContentPropertiesDialog(props) {
+	const theme = themeStyle(props.theme);
+	const textComps = [];
+	const [lines, setLines] = react_1.useState(0);
+	const [words, setWords] = react_1.useState(0);
```
Comment (laurent22) [inline ElectronClient/gui/NoteContentPropertiesDialog.js]: remove

File: ElectronClient/gui/MainScreen.jsx (hunk 9 above)
Comment (laurent22) [reply ElectronClient/gui/MainScreen.jsx]: Move to notebook

**Comment by laurent22** [comment]
We need something to show the hierarchy of notebooks though. Some people have hundreds of them and it wouldn't look good without indentation. Non-breaking spaces should work, especially since the prompt component is internal to Joplin and we aren't doing any filtering (or if we do, that should be disabled for non-breaking spaces).

Also you made your changes on a generated file instead of the TypeScript file. I don't understand what this NoteContentPropertiesDialog.js is even doing in your commit since it's in gitignore.

File: ElectronClient/gui/NoteContentPropertiesDialog.js (hunk 10 above)
Comment (coderrsid) [reply ElectronClient/gui/NoteContentPropertiesDialog.js]: Done.

**Comment by coderrsid** [comment]
Can you help me with the code.. i tried the same method you used in CliClipper. But it's not working i worked with some variations indeed with help of @tessus. But wasn't able to solve this..
image

**Comment by tessus** [comment]
@coderrsid you still haven't answered my question why you take screenshots of text instead of pasting the text in a code block.

**Comment by tessus** [comment]
also, the code you refer to is not in this PR

**Comment by laurent22** [comment]
@coderrsid, as tessus said, please post the code, not a screenshot. You ask for help with your code but if someone wants to try your code, how do they do? Just manually type everything based on your screenshot? Wouldn't that be simpler if you had just posted the code to begin with?

As for your question, I don't know the answer. I can help with Joplin-specific stuff but this is HTML/React in general, and probably you can find out by yourself.

If passing the options to `autocomplete` doesn't work, then find out how `autocomplete` is used. Perhaps it's trimming the data, but shouldn't? Or if it's some custom component that accepts HTML, maybe you can provide the data as an HTML string with `&nbsp;` to make the spaces, etc. This is the hard part of this PR but I assume it's not impossible to get it working, so please give it a try.

**Comment by tessus** [comment]
@coderrsid are you still working on this or shall we give this PR to someone else?

**Comment by coderrsid** [comment]
@tessus  I am still working on this PR.

**Comment by coderrsid** [comment]
@laurent22 I tried to find out the basic logic behind the prompt and as you said it's using a custom component `react-select` which takes all the different options from which user can select in `options` prop of the `Select` component.
I read the docs of react-select and for options it can be of only two type :
`type OptionType = { [string]: any }
type OptionsType = Array<OptionType>`
So for `react-select`, i don't think we'll be able to pass HTML for different options as you suggested. Instead what we can do is use JSX for creating the select instead of using it from third-party library `react-select`, only if you agree.

**Comment by tessus** [comment]
> i don't think we'll be able to pass HTML for different options as you suggested

Why? Have you tried it or was it explicitly stated in the documentation?

**Comment by coderrsid** [comment]
@tessus It's explicitly mentioned in the docs and my recent comment too as it allows only two types of options.

**Comment by tessus** [comment]
Well, `[string]: any` could also include HTML characters like `&nbsp;`. Your previous comment does not say anything about the string being processed literally.

**Comment by laurent22** [comment]
@coderrsid, you need to dig deeper here. The library doesn't support what you need? Then perhaps create a pull request on their repo to support it, or create a patch, or try a different component, etc. We can make suggestions on how to fix it, but you need to make some too, and try alternatives by yourself.

It needs to work with a notebook hierarchy, or if it's not possible we close the pull request. So please confirm if you're still interested looking into it.

**Comment by coderrsid** [comment]
@laurent22 Thanks for replying. Yeah sure I can dig deeper. It's just that i want us to stay on the same page. It shouldn't be like i spend some hours onto something which is not permissible and PackElend explicitly mentioned to communicate. That's why i am asking you guys for every step. Sorry if i am making it too much of a trouble. 😄

PS: I am working on this PR

**Comment by coderrsid** [comment]
I tried almost every method but unfortunately i guess the `react-select` doesn't support adding spaces to their options of `<Select  />` component. So i've created an issue in their repository. Link is below. You guys can also comment there, this would be great for us and their repo too. Thanks.
https://github.com/JedWatson/react-select/issues/3975

**Comment by PackElend** [comment]
> So i've created an issue in their repository. Link is below.

image
 😨

**Comment by coderrsid** [comment]
@PackElend  soo what else can we do?

**Comment by PackElend** [comment]
> @PackElend soo what else can we do?

is there an alternative library?

**Comment by coderrsid** [comment]
> is there an alternative library?
 yes, if we do try to find but instead of adding another library or package, can't. we render a JSX based select instead of using it from a library for just this case `Assign Notebook`?

**Comment by PackElend** [comment]
this not my call, I cannot help here.
Let's wait for the others 
@bedwardly-down I drag you in here too, you might be interested

**Comment by coderrsid** [comment]
Ohh sure @PackElend, i am happy for your participation.

**Comment by bedwardly-down** [comment]
I am going to be personally frank, I’ve been following this PR for a good while now and am exhausted of it. @coderrsid, if you can endure and follow through here, i will be impressed and amazed and would pretty much accept anything you want to do with Joplin for your project but also see that you failing would be a massive time sink for everyone involved here. I hope you get this completed and prove my fears and concerns wrong.

On the library question, I’m more than willing to help you figure that out but would also need to research it a bit because I don’t fully understand what all this PR is attempting to do just yet. I also know that many react native libraries seem to be getting abandoned as Facebook, its developer and maintainer, break compatibility through updates and whatnot.

**Comment by coderrsid** [comment]
>
>I hope you get this completed and prove my fears and concerns wrong.

Thanks @bedwardly-down . I will try my best for that. 

>
>I don’t fully understand what all this PR is attempting to do just yet.

We are just adding a functionality of assigning single or multiple notes to a notebook directly without the use of mouse-drag but using a dropdown like shown in the description of PR. 
As notebooks can contain sub-notebooks, we should use a tree hierarchy structure for listing notebook names in the dropdown like  :

tree

### instead of

non-tree

but the package `react-select` (used for rendering dropdown) doesn't support this and now we can try two different methods : 
- We can try a different package 
- We can render a React based dropdown without any package as used in CliClient above in the first GIF
- We can wait for https://github.com/JedWatson/react-select/issues/3975 to resolve

**Comment by laurent22** [comment]
@coderrsid, should we close this pull request as it's been nearly a month, many comments and people involved and it's still not done. Maybe it's too complicated to implement so it's perhaps time to cut our losses and close?

**Comment by coderrsid** [comment]
@laurent22 , please do not close this pull request, i am trying different libraries like `react-dropdown-tree-select` for tree type hierarchy. But will you accept adding another package to the application because the existing packages will not be able to pull up a tree hierarchy in the prompt-options dropdown?

**Comment by PackElend** [comment]
> @laurent22 , please do not close this pull request, i am trying different libraries like `react-dropdown-tree-select` for tree type hierarchy. 

Time is rare these days so you can continue working on it but Laurent won't react so please do not mention him anymore in this PR.
You can keep trying but you are on your own. You can ask in the forum for verification support as we need him and others elsewhere at the moment.

Your efforts and successful PRs have been recognized no worries :) 
May you can help out others too here and in the forum.

**Comment by coderrsid** [comment]
Yeah i know PackElend, sure i won't. I'll just try to implement it myself and i'll come with a solution very very soon. But please don't close this pull request. I've been working on this too along with everyone. 
Sure i'll help as much as i can. Thank you so much.

**Comment by Jackymancs4** [comment]
Hello,
I have been following this PR for some times now, mainly out of scientific interest. I think I found a viable solution that does not require a new package to be added.

Schermata 2020-03-25 alle 15 01 31
 
It leverages the styling API (https://react-select.com/styles) of `react-select` to adjust the left padding of a single option to a fixed amount.

The commit with the changes can be found here: https://github.com/Jackymancs4/joplin/commit/063a1649a0d2459ff62b746cb698163d52b52320 in https://github.com/Jackymancs4/joplin/tree/develop/assign-notebook

The code is just a mockup. To be functional this line https://github.com/Jackymancs4/joplin/blob/063a1649a0d2459ff62b746cb698163d52b52320/ElectronClient/gui/MainScreen.jsx#L215 needs to implement a proper logic to return the depth level of any notebook.

I hope this helps.

**Comment by coderrsid** [comment]
Hey @Jackymancs4 , thank you so much for figuring that out. You're right, it's about the styles in package. I tried implying the code by passing folder's indentDepth values in `MainScreen.jsx` as : 
```
else if (command.name === 'moveToFolder') {
			const startFolders = [];
			
            const addOptions = (folders, depth) => {
                for(let i = 0; i < folders.length; i++) {
                    const folder = folders[i];
                    startFolders.push({ key: folder.id, value: folder.id, label: folder.title, indentDepth: depth });
                    if(folder.children) addOptions(folder.children, depth + 1);
                }
            }

            addOptions(this.props.folders, 0);

			this.setState({
				promptOptions: {
					label: _('Move to notebook:'),
					inputType: 'dropdown',
					value: '',
					autocomplete: startFolders,
					onClose: async answer => {
						if (answer != null) {
							for (let i = 0; i < command.noteIds.length; i++) {
								await Note.moveToFolder(command.noteIds[i], answer.value);
							}
						}
						this.setState({ promptOptions: null });
					},
				},
			});
```
and the changes in `PromptDialog.jsx` are as : 
```
option: (provided, state) => {
				console.log(state.data);
				Object.assign(provided, {
					color: theme.color,
					fontFamily: theme.fontFamily,
					paddingLeft: `${10 + state.data.indentDepth * indentDepthWidth}px`,
				})},
```
But still when i `console.log` the `state.data` in the `PromptDialog.jsx`, it's passing `0` as the value for all folders as shown in the ss below.
image

I tried your way laurent as it is implied in Clipper. Can you tell why this is happening?

**Comment by Jackymancs4** [comment]
Well, this happen because `this.props.folders` does not contains `children` for any note (most likely it has been generated by `Folder.all()`).

In my last commit here https://github.com/laurent22/joplin/commit/979e8ec57aeff5f0d1baac4623a546541e422507 you can see a viable solution using `await Folder.allAsTree()`.

There I also added a mechanism to limit notebook depth, for good measure.
The result is fairly pleasant.

Schermata 2020-03-26 alle 13 33 49

**Comment by coderrsid** [comment]
Well, thanks for the viable solution but i just tried testing the changes, it renders them perfectly i'll just push the changes with some changes in styles. Thanks @Jackymancs4 for helping out big here!

**Comment by laurent22** [comment]
@coderrsid, there's some issue with the sorting order: https://github.com/laurent22/joplin/issues/3052 Do you know what might be the problem?

**Comment by coderrsid** [comment]
Sure laurent, i'll just look into it.
